Design.md

```diff
@@ -0,0 +1,149 @@
+# Design doc for job-runner
+
+Job-runner provides a gRCP API to execute arbitrary commands on a Linux host.
+
+## Runner
+
+The jobs themselves are normal child processes of the runner. The runner configures CGroups and Linux namespaces for children to limit resource usage and isolate the them from the other processes. It also handles output recording and forwarding. All data is stored in-memory. In case of the service process termination, all jobs and their status are lost.
+
+## Resource limits and isolation
+
+The job-runner uses cgroups V1 for resource limits. Cgroups will be used through an API wrapper like [controlgroup-rs](https://github.com/ordovicia/controlgroup-rs) or [cgroups-fs](https://github.com/frol/cgroups-fs). There are many tweakable options for limiting resource use, but the following minimal set has been selected to demonstrate how they work. In particular, the following cgroup parameters will be set:
+* [CPU](https://kernel.googlesource.com/pub/scm/linux/kernel/git/glommer/memcg/+/cpu_stat/Documentation/cgroups/cpu.txt):
+    * `cpu.cfs_period_us` (keep default: 100ms)
+    * `cpu.cfs_quota_us`
+        * `cfs_period_us` means one cpu cores worth of usage
+        * `cfs_period_us` / 2 means 50% of one core
+* [Memory](https://kernel.googlesource.com/pub/scm/linux/kernel/git/glommer/memcg/+/cpu_stat/Documentation/cgroups/memory.txt):
+    * `memory.limit_in_bytes` (max amount of memory to be used, hard limit)
+    * `memory.soft_limit_in_bytes` (set to 75% of `memory.limit_in_bytes`)
+    * `memory.memsw.limit_in_bytes` (set to `memory.limit_in_bytes`)
+* [Block IO](https://kernel.googlesource.com/pub/scm/linux/kernel/git/glommer/memcg/+/cpu_stat/Documentation/cgroups/blkio-controller.txt): (Device list readable from `/proc/partitions`)
+    * `blkio.throttle.read_bps_device`
+    * `blkio.throttle.write_bps_device`
+    * `blkio.throttle.read_iops_device`
+    * `blkio.throttle.write_iops_device`
+
+In addition to limiting resource use with cgroups, job-runner also isolates jobs from each other using namespaces. A PID namespace is set up to make sure the job cannot kill processes not spawned by it, and to make sure all child processes are terminated together with the actual job. A mount namespace is used to limit process to a subset of the file system, together with [`pivot_root(2)`](https://linux.die.net/man/2/pivot_root) (see [Understanding Containerization By Recreating Docker](https://itnext.io/linux-container-from-scratch-339c3ba0411d), search for pivot_root). Finally, a network namespace is created to limit network access of the jobs. It must be used together with [Virtual ethernet (VETH)](https://developers.redhat.com/blog/2018/10/22/introduction-to-linux-interfaces-for-virtual-networking#veth) if jobs should be allowed to communicate between each other or send data to the internet.
```

Comment:
Using a veth pair usually isn't the full story to allowing containerized networking. It's just one component of the setup. By limiting network access of the jobs what do you mean?

Reply:
That's true, usually you need to create bridge networks by connecting via veths to bridge interfaces.

By limiting network access I simply mean that by default no network adapters are not available for processes in network namespaces. They cannot communicate with the network until given access to it manually. Of course it would be possible to setup firewall rules on the interface, but I think that would be out-of-scope.